zerver/tests/test_custom_profile_data.py

```diff
@@ -454,6 +454,69 @@ def test_delete_internals(self) -> None:
         self.assertFalse(self.custom_field_exists_in_realm(field.id))
         self.assertEqual(user_profile.customprofilefieldvalue_set.count(), self.original_count - 1)
 
+    def test_delete_value_with_editable_by_user(self) -> None:
+        iago = self.example_user("iago")
+        hamlet = self.example_user("hamlet")
+        realm = iago.realm
+        self.login("hamlet")
+
+        biography_custom_field = CustomProfileField.objects.get(name="Biography", realm=realm)
+        birthday_custom_field = CustomProfileField.objects.get(name="Birthday", realm=realm)
+
+        # Set and assert our initial state.
+        data = {}
+        data["editable_by_user"] = "false"
+        result = self.api_patch(
+            iago, f"/api/v1/realm/profile_fields/{birthday_custom_field.id}", info=data
+        )
+        self.assert_json_success(result)
+
+        birthday_custom_field.refresh_from_db()
+        self.assertFalse(birthday_custom_field.editable_by_user)
+        self.assertTrue(biography_custom_field.editable_by_user)
+
+        CustomProfileFieldValue.objects.get(user_profile=iago, field=birthday_custom_field)
+        CustomProfileFieldValue.objects.get(user_profile=hamlet, field=birthday_custom_field)
+        CustomProfileFieldValue.objects.get(user_profile=iago, field=biography_custom_field)
+        CustomProfileFieldValue.objects.get(user_profile=hamlet, field=biography_custom_field)
+
+        # Users can only delete fields where editable_by_user is true.
+        result = self.client_delete(
+            "/json/users/me/profile_data",
+            {"data": orjson.dumps([biography_custom_field.id]).decode()},
+        )
+        self.assert_json_success(result)
+        self.assertFalse(
+            CustomProfileFieldValue.objects.filter(
+                user_profile=hamlet, field=biography_custom_field
+            ).exists()
+        )
+
+        result = self.client_delete(
+            "/json/users/me/profile_data",
+            {"data": orjson.dumps([birthday_custom_field.id]).decode()},
+        )
+        self.assert_json_error(result, "User does not have permission to update the field.")
+        self.assertTrue(
+            CustomProfileFieldValue.objects.filter(
+                user_profile=hamlet, field=birthday_custom_field
+            ).exists()
+        )
+
+        # Admins can always delete fields for themselves and other users regardless of
+        # editable_by_user. We test them deleting for their own profile here.
```

Comment:
I see that the `DELETE` endpoint is only for user updating the field value for themselves and not others. So, might be just better to just not mention the "other users" case in the comment here as we are not testing that.

Reply:
Updated the comment here.